### PR TITLE
feat(ip-restriction): Add TCP Support

### LINKED ISSUE #6679: IP Restriction at TCP Level.
### Summary
I am using Kong IP restriction plugin, according to the documentation its only compatible with HTTP and HTTPs, wanted to apply it at TCP level. Is there such a provision in Kong? Or a work around.

### PR DESCRIPTION
**Pull request by scrudge**
### Summary
This change adds TCP support to the ip-restriction plugin by implementing the Stream module's preread function.

When a TCP connection is rejected due to IP restriction rules, a JSON error response is written to the stream and the connection is closed.

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [x] Link to docs pr: https://github.com/Kong/docs.konghq.com/pull/5736

### Full changelog
* Add protocols to schema: "tcp", "tls", "grpc", "grpcs"
* Implement preread function
* Make handler logic generic to be shared by HTTP and TCP

### Issue reference

Fix #6679 

KAG-722

### CODE REVIEW

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:1]: "cjson.safe" may be better.

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:66]: We should localize `cjson.encode` at the module level.

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:88]: localize `string.format`, and `binary_remote_addr` may not be formated as `%s`.

**Comment by chronolaw** [comment]
This PR is familiar with #10229, What is the relation?

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:88]: We work on the same team.  We need this feature in 2.8.x because we can't upgrade to v3 yet.

**Comment by scrudge** [comment]
> 

We work on the same team.  We need this feature in 2.8.x because we can't upgrade to v3 yet.

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:88]: I'm not sure what you mean by localize.  Put the value in a local variable first?

Comment (chronolaw) [reply kong/plugins/ip-restriction/handler.lua:88]: like `local string_format = string.format`.

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: Done

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:56]: I think we can remove this arg `is_http`.

In `stream` subsystem we will call `preread()`, and in `http` subsystem we will call `access(()`.

So we could do this to eliminate the runtime cost:
```
local is_http_subsystem = ngx.config.subsystem == "http"
local do_exit
if is_http_subsystem then
  do_exit = function() ... end
else
  do_exit = function() ... end
end
```

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:61]: Same as others, localize `ngx.req.socket`

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [reply kong/plugins/ip-restriction/handler.lua]: I think you misunderstand my meaning. The code should be:
```
local cjson_encode = cjson.encode
```

**Comment by chronolaw** [comment]
Could you add some test cases for this PR?  Thanks.

**Comment by scrudge** [comment]
> Could you add some test cases for this PR? Thanks.

Sure, but I'm completely lost on how testing works here.  is this the test for this plugin :
https://github.com/Kong/kong/blob/master/t/01-pdk/03-ip/01-is_trusted.t

Also, is there some docs on how testing works? Thanks.

**Comment by chronolaw** [comment]
This plugin's test should be `spec/03-plugins/17-ip-restriction`.
Perhaps we do not have a proper doc for the test, but don't worry, we can implement the feature first.

**Comment by scrudge** [comment]
Yeah, I found the tests right after I asked :-)

I'm actually working on the other suggested changed ATM.
I'll also try to fix the test that are broken from the change in response message.

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:44]: Move this line out of function

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:42]: blank line before `else`

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [inline kong/plugins/ip-restriction/handler.lua:94]: We can not output `binary_remote_addr` as `%s`, perhaps we should use the original message `Your IP address is not allowed`

File: kong/plugins/ip-restriction/handler.lua
Comment (sabertobihwy) [inline kong/plugins/ip-restriction/handler.lua:82]: I think change the func name `handler` to `do_restrict` is better.

File: kong/plugins/ip-restriction/handler.lua
Comment (sabertobihwy) [inline kong/plugins/ip-restriction/handler.lua:55]: 1. It is better to localize `ngx.exit`

```lua
local ngx_exit = ngx.exit

...

ngx_exit()
```

2. we should use `ngx.exit()` with a status argument, e.g. `ngx.exit(1)`

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: Sorry, we had fixed this in our code and I forgot to copy it over.  I changed it to the string address variable.  We would really like to keep this feature.  It's been very useful when trouble shooting with users and identifying holes in the allow lists.

File: kong/plugins/ip-restriction/handler.lua
Comment (sabertobihwy) [inline kong/plugins/ip-restriction/handler.lua:57]: sorry, I misled you, maybe `ngx_exit(status)` would be better here, right? cc @dndx

File: kong/plugins/ip-restriction/handler.lua
Comment (hbagdi) [inline kong/plugins/ip-restriction/handler.lua:54]: Why send a 'status' property for TCP/UDP connections? HTTP status is an L7 concept and we should not be sending that to a client who is using TCP/UDP.

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:54]: I understand your point.  What type of response would you recommend?

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:54]: @hbagdi Any suggestion on the response format and content?

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: @chronolaw is the fix acceptable?

File: kong/plugins/ip-restriction/handler.lua
Comment (hbagdi) [reply kong/plugins/ip-restriction/handler.lua:54]: `{"message":"rate-limit exceeded"}` would make more sense.

Comment (merusso) [reply kong/plugins/ip-restriction/handler.lua:54]: So would the following be acceptable?

```lua
    tcpsock:send(cjson_encode({
      message = message
    }))
```

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:53]: Status removed

**Comment by hanshuebner** [comment]
@scrudge Did you mean to add a test for the new functionality?  I only see changes to the existing test.  Please rebase to the `master` branch as well.  Thanks!

**Review by hanshuebner — CHANGES_REQUESTED** [review]
Missing test for new functionality.

File: kong/plugins/ip-restriction/handler.lua
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua:56]: Could this just use `kong.response.exit`?

```suggestion
    kong.response.exit(status, { message = message })
```

File: kong/plugins/ip-restriction/handler.lua
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua:89]:
```suggestion
    return do_exit(status, message)
```

File: kong/plugins/ip-restriction/handler.lua
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua:101]:
```suggestion
      return do_exit(status, message)
```

File: kong/plugins/ip-restriction/handler.lua
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua:108]:
```suggestion
      return do_exit(status, message)
```

**Review by bungle — DISMISSED** [review]
Could you please add some return statements (it makes code cleaner about its intent).

**Comment by hbagdi** [comment]
@scrudge Could you please address the latest feedback?

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:56]: Can kong.response.exit(status) work with the TCP connection?

**Comment by scrudge** [comment]
> Missing test for new functionality.

As I mentioned early, I have know idea how to test TCP.  Is there some example you can suggest.  Thanks.

File: kong/plugins/ip-restriction/handler.lua
Comment (chronolaw) [reply kong/plugins/ip-restriction/handler.lua:56]: Yes, see `pdk/response.lua`

File: spec/03-plugins/17-ip-restriction/02-access_spec.lua
Comment (chronolaw) [inline spec/03-plugins/17-ip-restriction/02-access_spec.lua:251]: We could use `assert.matches()` with the special pattern.

**Comment by hbagdi** [comment]
@chronolaw How do we test this feature? Could you please share some thoughts on this here?

**Comment by chronolaw** [comment]
I think we could find some example code in `spec/02-integration/05-proxy`, such as `28-stream_plugins_triggering_spec.lua` and `19-grpc_proxy_spec.lua`.

**Comment by scrudge** [comment]
Yes, thanks.

**Comment by jjchambl** [comment]
Who do I need to yell at to reopen this

**Comment by scrudge** [comment]
@hbagdi can we re-open this?  We are about to update with some tests.

**Comment by jeffrey-arndt** [comment]
Reopening. Tests to be added soon

File: spec/03-plugins/17-ip-restriction/02-access_spec.lua
Comment (scrudge) [reply spec/03-plugins/17-ip-restriction/02-access_spec.lua:309]: Can you send a link to the docs for assert.matches()?  I can't find it.

**Comment by scrudge** [comment]
Sorry, I tried applying a last minute suggestion and it broke the build.
I've rolled it back and added tests, which are all passing.  I believe this is ready for review.

File: kong/plugins/ip-restriction/handler.lua
Comment (hanshuebner) [reply kong/plugins/ip-restriction/handler.lua:56]: Seems like this is not addressed.

**Review by hanshuebner — CHANGES_REQUESTED** [review]
One remaining comment to be addressed, otherwise good to go.  I have added a [docs PR](https://github.com/Kong/docs.konghq.com/pull/5736) as well.

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:56]: The Kong response caused an error when I tested it.
However, I'll remove the json response that should make it more efficient.

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: I removed the json response.

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: @hanshuebner is this acceptable?

Comment (hanshuebner) [reply kong/plugins/ip-restriction/handler.lua]: I think @bungle pointed out that `kong.response.exit` should be preferred to `ngx.exit`.  Any reason not to follow his advice?

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: Yes, I tried using kong.response.exit and it failed sending the reponse.

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: @hanshuebner can we close this?  I went with one of @bungle suggestions.  The other didn't work.

**Comment by scrudge** [comment]
@hanshuebner @bungle 

There's one remaining comment.  I think I've resolved the issue, just waiting for feedback.
https://github.com/Kong/kong/pull/10245#discussion_r1123420637

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: It's not sending the message.  I'm seeing an empty string:
https://github.com/scrudge/kong/actions/runs/5391318477/jobs/9787991403#step:9:1342

**Comment by hbagdi** [comment]
Please add an entry to CHANGELOG.md.

File: kong/plugins/ip-restriction/handler.lua
Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:50]: We use that for debugging, But I can see issues with protocols like ssh.
Getting rid of the message would also break all of the tests we added.
@hbagdi how would you like us to proceed?

Comment (hbagdi) [reply kong/plugins/ip-restriction/handler.lua:50]: @chronolaw @hanshuebner Opinions?
Why not only close the socket and do nothing more?
(please hold until we arrive at a solution)

Comment (hanshuebner) [reply kong/plugins/ip-restriction/handler.lua:50]: @hbagdi I agree that sending a message through the TCP channel is unhygienic.  Furthermore, one could argue that from a security perspective, it may be undesirable to reveal the fact that there is a black- or whitelist that causes the closure of the connection.  From a usability perspective, however, I feel that sending a message before closing the connection is very helpful in debugging.  Making the message sending configurable would be an option that'd also help with testability.  Another option will be to change the tests so that instead of looking for the in-band rejection message, they will look at the log file.

Personally, I'll be fine keeping things how they are.

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua:50]: I vote for keeping the message, but am happy to do whatever.  @hbagdi who is going to make the final call?  Please let me know how to proceed.  Would like like to wrap this up, we've been working on this for quite a while and are currently having to use a custom build for now.  Thanks.

Comment (hbagdi) [reply kong/plugins/ip-restriction/handler.lua:50]: Mixing up strings on an arbitrary TCP connection doesn't seem the right thing to do unless someone can prove enough value for doing the wrong thing.

@hanshuebner Could you please check if Kong does something similar on a TCP connection either in the core or plugin? I don't recall anything like that.
If this is a net new thing, let's remove the message and merge this. If a message is needed, we will do an opt-in flag on the plugin configuration for it.

Comment (hanshuebner) [reply kong/plugins/ip-restriction/handler.lua:50]: I could not find any evidence that we do such in-band error signaling on TCP sockets at this point.

@scrudge Can you remove the message and update your tests so that they look at specific log entries instead?  This is done in several other tests as well, here's an example: https://github.com/Kong/kong/blob/96eb92f2e301b4b249eae710d51715931a683656/spec/02-integration/05-proxy/06-ssl_spec.lua#L270

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: I refactored to use kong respnse error (which does not return the message on TCP connections.  Also added a log message at level warn when a connection is denied.  Updated the test to use the log message to validate the deny.

Comment (scrudge) [reply kong/plugins/ip-restriction/handler.lua]: @hanshuebner @hbagdi 
Do you mind reviewing my changes?
Thanks!

**Review by hanshuebner — APPROVED** [review]

**Comment by scrudge** [comment]
> Could you please add some return statements (it makes code cleaner about its intent).

@bungle can you please approve changes you requested?  Thank you.

**Review by merusso — APPROVED** [review]

**Comment by scrudge** [comment]
> > Could you please add some return statements (it makes code cleaner about its intent).
> 
> @bungle can you please approve changes you requested? Thank you.

@hanshuebner @hbagdi 
I believe this PR is held waiting for a response from @bungle but I haven't heard from him since his original comment.  Is there anyway to get this completed?
Thanks for all the help.

**Comment by hanshuebner** [comment]
@scrudge Thank your for your contribution!

**Comment by scrudge** [comment]
Thank you for all your help.

**Pull request merged**
